### PR TITLE
[Api] Tests adding incorrect county code to address order

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | 1.9
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: features/cart/shopping_cart/allowing_access_only_for_correctly_logged_in_users.feature [hunk 1]
```diff
@@ -189,3 +189,11 @@ Feature: Allowing access only for correctly logged in users
...
         When the administrator try to see the summary of customer's cart
         Then the administrator should see "Stark T-shirt" product with quantity 1 in the visitor cart
+
+    @api
+    Scenario: Trying to add incorrect address to the cart by the visitor
+        Given the visitor has product "Stark T-Shirt" in the cart
+        When the visitor specify the email as "jon.snow@example.com"
+        And the visitor try to specify the billing incorrect address as "Ankh Morpork", "Frost Alley", "90210", "United Russia" for "Jon Snow"
+        And the visitor complete the addressing step
+        Then I should be notified that "United Russia" country does not exist
```
Comment (GSadee) [inline features/cart/shopping_cart/allowing_access_only_for_correctly_logged_in_users.feature:199]: This scenario doesn't fit this feature. You should move probably to the new one in `features/checkout/addressing_order` directory. What is more, you should cover also the case, when you are providing an address without country code defined.

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectAddressOrder.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+final class CorrectAddressOrder extends Constraint
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectAddressOrder.php:18]:
```suggestion
final class CorrectOrderAddress extends Constraint
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectAddressOrderValidatorSpec.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    function it_is_a_constraint_validator(): void
+    {
+        $this->shouldImplement(ConstraintValidatorInterface::class);
+    }
+
+    function it_throws_an_exception_if_value_is_not_an_instance_of_address_order_command(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [new CompleteOrder(), new AddingEligibleProductVariantToCart()]);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectAddressOrderValidatorSpec.php:39]: You should pass here a correct constraint
```suggestion
            ->during('validate', [new CompleteOrder(), new CorrectAddressOrder()]);
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectAddressOrderValidatorSpec.php [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    function it_throws_an_exception_if_constraint_is_not_an_instance_of_adding_eligible_product_variant_to_cart(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [
+                new AddItemToCart('productCode', 'productVariantCode', 1),
+                new class() extends Constraint {}
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectAddressOrderValidatorSpec.php:48]: You should pass here a correct command:
```suggestion
                new AddressOrder(...),
                new class() extends Constraint {}
```

File: features/checkout/addressing_order/order_address_validation.feature [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        Given the store operates on a single channel in "United States"
+        And the store has a product "Stark T-shirt" priced at "$12.00"
+        And the store allows paying offline
+        And the store has "UPS" shipping method with "$20.00" fee
+
+    @api
+    Scenario: Trying to add address with incorrect country to the cart by the visitor
+        Given the visitor has product "Stark T-Shirt" in the cart
+        When the visitor specify the email as "jon.snow@example.com"
+        And the visitor try to specify the billing incorrect address as "Ankh Morpork", "Frost Alley", "90210", "United Russia" for "Jon Snow"
```
Comment (Zales0123) [inline features/checkout/addressing_order/order_address_validation.feature:17]: `the billing incorrect address` -> `the incorrect billing address`

File: features/checkout/addressing_order/order_address_validation.feature [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        And the visitor complete the addressing step
+        Then I should be notified that "United Russia" country does not exist
+
+    @api
+    Scenario: Trying to add address without country to the cart by the visitor
+        Given the visitor has product "Stark T-Shirt" in the cart
+        When the visitor specify the email as "jon.snow@example.com"
+        And the visitor try to specify the billing address without country as "Ankh Morpork", "Frost Alley", "90210" for "Jon Snow"
```
Comment (Zales0123) [inline features/checkout/addressing_order/order_address_validation.feature:25]: Do we need this `without country`? It's quite visible there is no country specified :smile:

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 7]
```diff
@@ -619,6 +648,28 @@ public function theVisitorShouldHaveCheckoutAddressStepCompleted(string $stepTyp
         Assert::same($this->getCheckoutState(), $this::CHECKOUT_STATE_TYPES[$stepType]);
     }
 
+    /**
+     * @Then I should be notified that :countryName country does not exist
+     */
+    public function iShouldNotBeNotifiedThatCountryDoesNotExist(string $countryName): void
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:654]:
```suggestion
    public function iShouldBeNotifiedThatCountryDoesNotExist(string $countryName): void
```

File: features/checkout/addressing_order/order_address_validation.feature [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        And the visitor complete the addressing step
+        Then I should be notified that address without country does not exist
```
Comment (Zales0123) [inline features/checkout/addressing_order/order_address_validation.feature:27]:
```suggestion
        Then I should be notified that the country is required to address
```

or 

```suggestion
        Then I should be notified that address without country cannot exist
```

?

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 9]
```diff
@@ -1261,4 +1312,20 @@ private function completeOrder(): Response
 
         return $this->ordersClient->executeCustomRequest($request);
     }
+
+    private function addAddressWithoutCountry(
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:1316]: More like `formatAddressRequestContent` with an optional argument `?string $countryCode`?

File: src/Sylius/Bundle/ApiBundle/Resources/translations/validators.en.yaml [hunk 10]
```diff
@@ -1,4 +1,8 @@
 sylius:
+    address:
+        without_country: 'The address without country can not exist'
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Resources/translations/validators.en.yaml:3]: `can not` -> `cannot`?

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+use Sylius\Bundle\ApiBundle\Command\Checkout\AddressOrder;
+use Sylius\Component\Core\Model\AddressInterface;
+use Symfony\Component\Intl\Countries;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+final class CorrectOrderAddressValidator extends ConstraintValidator
+{
+    public function validate($value, Constraint $constraint)
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php:25]: `: void` 💃

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+        if ($countryCode === null) {
+            $this->context->addViolation(
+                $constraint->addressWithoutCountryCodeCanNotExistMessage
+            );
+
+            return;
+        }
+
+        if (!Countries::exists($countryCode)) {
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php:54]: I believe we should use the `countryRepository` and check if the country exists within Sylius

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    private function validateAddress(?AddressInterface $address, CorrectOrderAddress $constraint): void
+    {
+        if ($address === null) {
+            return;
+        }
+
+        /** @var string|null $countryCode */
+        $countryCode = $address->getCountryCode();
+
+        if ($countryCode === null) {
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php:46]: I'm wondering, isn't it the case for the other validator 😄 We would then have two constraints e.g. `CountrySpecified` and `CountryExisting`? Seems as a better separation of responsibilities

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectOrderAddressValidatorSpec.php [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    function it_is_a_constraint_validator(): void
+    {
+        $this->shouldImplement(ConstraintValidatorInterface::class);
+    }
+
+    function it_throws_an_exception_if_value_is_not_an_instance_of_address_order_command(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [new CompleteOrder(), new CorrectOrderAddress()]);
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectOrderAddressValidatorSpec.php:36]: Semicolon should be in the next line 💃

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectOrderAddressValidatorSpec.php [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        AddressInterface $shippingAddress
+    ): void {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [
+                new AddressOrder(
+                    'john@doe.com',
+                    $billingAddress->getWrappedObject(),
+                    $shippingAddress->getWrappedObject()
+                ),
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/CorrectOrderAddressValidatorSpec.php:50]: Wouldn't it fit in one line?

File: features/checkout/addressing_order/order_address_validation.feature (hunk 6 above)
Comment (Tomanhez) [reply features/checkout/addressing_order/order_address_validation.feature:25]: It create more clearer this step IMO :)

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php (hunk 13 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Validator/Constraints/CorrectOrderAddressValidator.php:55]: Its possible, but IMO it is just `addressValidator`. I can extract it to other services, but is it too much code to simple address validation? :D And we can add here next points for other address fields.

**Review by AdamKasp — APPROVED** [review]

File: features/checkout/addressing_order/order_address_validation.feature [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+        And the store has a product "Stark T-shirt" priced at "$12.00"
+        And the store allows paying offline
+        And the store has "UPS" shipping method with "$20.00" fee
+
+    @api
+    Scenario: Trying to add address with incorrect country to the cart by the visitor
+        Given the visitor has product "Stark T-Shirt" in the cart
+        When the visitor specify the email as "jon.snow@example.com"
+        And the visitor try to specify the incorrect billing address as "Ankh Morpork", "Frost Alley", "90210", "United Russia" for "Jon Snow"
+        And the visitor complete the addressing step
```
Comment (GSadee) [inline features/checkout/addressing_order/order_address_validation.feature:18]: These steps can be merged into one, the second one is not needed for API context

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 17]
```diff
@@ -182,6 +182,33 @@ public function iSpecifyTheBillingAddressAs(AddressInterface $address): void
...
+    public function iTryToSpecifyTheIncorrectBillingAddressAs(
+        string $city,
+        string $street,
+        string $postcode,
+        string $countryName,
+        string $customerName
+    ): void {
+        $addressType = 'billingAddress';
+
+        $this->addAddress($addressType, $city, $street, $postcode, $customerName, $countryName);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:197]:
```suggestion
        $this->addAddress('billingAddress', $city, $street, $postcode, $customerName, $countryName);
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 18]
```diff
@@ -182,6 +182,33 @@ public function iSpecifyTheBillingAddressAs(AddressInterface $address): void
...
+    }
+
+    /**
+     * @When /^the visitor try to specify the billing address without country as "([^"]+)", "([^"]+)", "([^"]+)" for "([^"]+)"$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:201]: You can name the arguments and then merge this method with the one above
```suggestion
     * @When the visitor try to specify the billing address without country as :city, :street, :postcode for :customerName
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 19]
```diff
@@ -182,6 +182,33 @@ public function iSpecifyTheBillingAddressAs(AddressInterface $address): void
         $this->fillAddress('billingAddress', $address);
     }
 
+    /**
+     * @When /^the visitor try to specify the incorrect billing address as "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)" for "([^"]+)"$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:186]:
```suggestion
     * @When the visitor try to specify the incorrect billing address as :city, :street, :postcode, :countryName for :customerName
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Tomasz! :tada:

**Pull request merged**